### PR TITLE
HeaderParsing README

### PR DESCRIPTION
**Pull request by sebastienros**
null

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/extensions/pull/4528)

### CODE REVIEW

File: src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md [hunk 1]
```diff
@@ -1 +1,130 @@
...
+
+```dotnetcli
+dotnet add package Microsoft.AspNetCore.HeaderParsing
+```
+
+Or directly in the C# project file:
+
+```xml
+<ItemGroup>
+  <PackageReference Include="Microsoft.AspNetCore.HeaderParsing" Version="*" />
```
Comment (Tratcher) [inline src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md:25]: Don't encourage bad habits
```suggestion
  <PackageReference Include="Microsoft.AspNetCore.HeaderParsing" Version="[CURRENTVERSION]" />
```

File: src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md [hunk 2]
```diff
@@ -1 +1,130 @@
...
+```
+## Usage Example
+
+### Registering the services
+
+The services can be registered using one of the `AddHeaderParsing` overloads:
+
+```csharp
+public static IServiceCollection AddHeaderParsing(this IServiceCollection services);
+public static IServiceCollection AddHeaderParsing(this IServiceCollection services, Action<HeaderParsingOptions> configuration);
```
Comment (Tratcher) [inline src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md:36]: Shouldn't this be 
```suggestion
public static IServiceCollection AddHeaderParsing(this IServiceCollection services, Action<HeaderParsingOptions> configure);
```

File: src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md (hunk 1 above)
Comment (sebastienros) [reply src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md:25]: I had no better idea :/ I noticed that @joperezr cheated and only mentioned the dotnet-cli way ;) I think we should converge and use the same pattern everywhere. At least this starts the discussion.

File: src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md (hunk 2 above)
Comment (sebastienros) [reply src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md:36]: Thanks, I will report it to the API review. And the other instances I found BTW.

Comment (sebastienros) [reply src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md:36]: I am merging this suggestion, but the API is wrong, this will need to be fixed.

Comment (Tratcher) [reply src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md]: Want to just fix the API in this PR?

Comment (sebastienros) [reply src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md]: Done

File: src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md [hunk 3]
```diff
@@ -1 +1,130 @@
...
+
+In particular:
+
+- There are many different headers with subtle format differences that are hard to parse correctly in all cases. This package provides predefined parsers for commonly used headers.
+- Your application may parse the same header value multiple times per request. This package automatically caches the parsed values of these headers to make your application faster.
+- This package also provides logging and metering of bad header values.
+
+## Getting started
+
+### Install the package
```
Comment (Tratcher) [inline src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md:13]: Seems redundant?
```suggestion
## Install the package
```

Comment (sebastienros) [reply src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md:13]: Followed @joperezr's README for consistency.

Comment (Tratcher) [reply src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md:13]: Where?

Comment (sebastienros) [reply src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md:13]: https://github.com/dotnet/extensions/pull/4512/files#diff-358a8dab66534c16517c7dc752a9f3926e6a3b4b3c2b713cd32d908d8be1f73bR5-R7

File: src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md [hunk 4]
```diff
@@ -1 +1,128 @@
-README
+# Microsoft.AspNetCore.HeaderParsing
+
+This package provides services to for strongly typed header parsing and value caching.
+
+In particular:
+
+- There are many different headers with subtle format differences that are hard to parse correctly in all cases. This package provides predefined parsers for commonly used headers.
+- Your application may parse the same header value multiple times per request. This package automatically caches the parsed values of these headers to make your application faster.
```
Comment (Tratcher) [inline src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md:8]:
```suggestion
- Your application may parse the same header value multiple times per request. These APIs automatically cache the parsed values of these headers to make your application faster.
```

File: src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md [hunk 5]
```diff
@@ -1 +1,128 @@
...
+- [IfUnmodifiedSince](https://learn.microsoft.com/dotnet/api/microsoft.net.http.headers.headernames.IfUnmodifiedSince)
+- [Range](https://learn.microsoft.com/dotnet/api/microsoft.net.http.headers.headernames.Range)
+- [Referer](https://learn.microsoft.com/dotnet/api/microsoft.net.http.headers.headernames.Referer)
+- XForwardedFor (`X-Forwarded-For`)
+
+## Options
+
+The `HeaderParsingOptions` class is used to configure common behaviors for this feature.
+
+- In case you configured parsing of a specific header, but the header is not present in the current HTTP request, you can pre-configure default values to be returned in such cases. See the `DefaultValues` option.
```
Comment (danmoseley) [inline src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md:106]:
```suggestion
- You can pre-configure default values to be returned when the current HTTP request contains headers for which you have not configured parsing. See the `DefaultValues` option.
```
?

File: src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md [hunk 6]
```diff
@@ -1 +1,128 @@
...
+- The number of cached values can be changed per header using the `MaxCachedValuesPerHeader` option. If not configured the `DefaultMaxCachedValuesPerHeader` value will be used.
```
Comment (danmoseley) [inline src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md:107]: nit, you're mixing active ("you can") and passive ("can be")

```suggestion
- You can change the number of cached values using the `MaxCachedValuesPerHeader` option. If not configured the `DefaultMaxCachedValuesPerHeader` value will be used.
```

File: src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md [hunk 7]
```diff
@@ -1 +1,128 @@
...
+
+## Custom headers parsing
+
+Although this package provides support for parsing many HTTP headers, you sometimes need to parse custom headers.
+You can easily extend the header parsing model to support your custom headers by implementing the abstract class `HeaderParser` and then register an 
+instance of this class in the `IHeaderRegistry` interface.
+
+## Feedback & Contributing
+
+For any feedback or contributions, please visit us in [our GitHub repo](https://github.com/dotnet/extensions).
```
Comment (danmoseley) [inline src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md:128]: I kind of liked the "We welcome.." wording I suggested in Chris's PR, because it actively encourages folks to engage not just telling them where ...

Comment (sebastienros) [reply src/Libraries/Microsoft.AspNetCore.HeaderParsing/README.md:128]:
```suggestion
We welcome feedback and contributions in [our GitHub repo](https://github.com/dotnet/extensions).
```

**Review by Tratcher — APPROVED** [review]
